qutip/random_objects.py

```diff
@@ -65,115 +143,154 @@ def rand_jacobi_rotation(A, *, seed=None):
-    The Ginibre ensemble corresponds to setting ``norm = 1`` [Mis12]_.
+    Obtain a list of matrices block sizes in a way that a NxN matrix composed.
+    Of full matrices of these size allong the diagonal would be of desired
+    density.
+    """
+    if density <= 0:
+        return [1] * N
+    elif density >= 1:
+        return [N]
+    max_block_size = int((N**2 * density)**0.5)
+    min_block_size = 1
+    if density >= 0.5:
+        min_block_size = int(N / 2 * (1 + (2 * density - 1)**0.5))
+
+    if min_block_size >= max_block_size:
+        M = min(min_block_size, max_block_size)
+    else:
+        M = generator.integers(min_block_size, max_block_size) + 1
+    M += generator.integers(-1, 2)
+    M = min(max(M, 1), N)
 
-    Parameters
-    ----------
-    shape : tuple
-        Shape of the returned array of random variates.
+    other_blocks = []
+    if M < N:
+        other_N = N - M
+        other_density = (density * N**2 - M**2) / other_N**2
+        other_blocks = _get_block_sizes(other_N, other_density, generator)
+
+    return [M] + other_blocks
 
-    norm : float
-        Scale of the returned random variates, or 'ginibre' to draw
-        from the Ginibre ensemble.
 
-    seed : int32
-        seed to reseed the random number generator.
+def _merge_shuffle_blocks(blocks, generator):
     """
-    # This function is intended for internal use.
-    if seed is not None:
-        np.random.seed(seed=seed)
-    if norm == 'ginibre':
-        norm = 1
-    return np.sum(np.random.randn(*(shape + (2,))) * _UNITS, axis=-1) * norm
+    For a list of block, merge them in one matrix along the diagonal and
+    shuffle the rows and columns.
+    """
+    N = sum(block.shape[0] for block in blocks)
+    D = sum(block.shape[0]**2 for block in blocks) / N**2
+    idx = generator.permutation(N)
+    end = 0
+    if D < 0.1:
+        # TODO: coo_array from sicpy 1.8 would allow to simplify this.
+        row = []
+        col = []
+        data = []
+        for block in blocks:
+            start = end
+            end = end + block.shape[0]
+            brow, bcol = np.meshgrid(idx[start:end], idx[start:end])
+            row.append(brow.ravel())
+            col.append(bcol.ravel())
+            data.append(block.ravel())
+        data = np.hstack(data)
+        row = np.hstack(row)
+        col = np.hstack(col)
+        matrix = sp.coo_matrix((data, (row, col)), shape=(N, N))
+    else:
+        matrix = np.zeros((N,N), dtype=complex)
+        for block in blocks:
+            start = end
+            end = end + block.shape[0]
+            brow, bcol = np.meshgrid(idx[start:end], idx[start:end])
+            matrix[brow, bcol] = block
+    return _data.create(matrix, copy=False)
 
 
-def rand_herm(N, density=0.75, dims=None, pos_def=False,
-              *, seed=None, dtype=_data.CSR):
-    """Creates a random NxN sparse Hermitian quantum object.
+def rand_herm(dimensions, density=0.30, pos_def=False, eigenvalues=None, *,
```

Comment:
Why was the default changed from 0.75 to 0.30?

Reply:
I forgot I changed that. Do you prefer the previous value?
We often use this for a random Hamiltonian, since they are usual usually quite sparse, this seems a more useful value.

Reply:
I'm personally happy with a sparser value, but I'm wondering if it will affect users. Probably not, since it's already random in any case.